src/main/frontend/filtering.cljs

```diff
@@ -0,0 +1,43 @@
+(ns frontend.filtering
+  (:require [frontend.util :as util]
+            [frontend.state :as state]
+            [frontend.db :as db]))
+
+(defn get-block-references
+  [block]
+  (map #(if (= (first %) "Tag") (second %) (second (:url (second %))))
+       (filter #(and (some (partial = (first %)) ["Tag" "Link"])
+                     (or (= (first %) "Tag")
+                         (= (first (:url (second %))) "Search")))
+               (:block/title block))))
+
+(defn get-nested-block-references
+  [block]
+  (let [repo (state/get-current-repo)
+        children (-> (db/get-block-immediate-children repo (:block/uuid block))
+                     db/sort-by-pos)
+        child-refs (map get-nested-block-references children)
+        block-refs (get-block-references block)]
+    (flatten (concat child-refs block-refs))))
+
+(defn matches-filter
+  [references filter-state]
+  (every? #(= (util/in? (first %) references) (second %)) filter-state))
+
+(defn block-matches-filter
+  [block filter-state]
+  (let [repo (state/get-current-repo)
+        children (-> (db/get-block-immediate-children repo (:block/uuid block))
+                     db/sort-by-pos)
+        block-matches (matches-filter (get-block-references block) filter-state)
+        child-matches (delay (some #(block-matches-filter % filter-state) children))]
+    (or block-matches @child-matches)))
```

Comment:
The code could be rewritten as below to avoid the `delay` fn:
```
(or block-matches 
   (some #(block-matches-filter % filter-state) children))
```
We can introduce a let binding if we want to make the code easier to read.